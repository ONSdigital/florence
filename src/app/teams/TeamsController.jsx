import React, { Component } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import PropTypes from 'prop-types';

import { 
    updateAllTeams, 
    updateActiveTeam, 
    emptyActiveTeam, 
    updateActiveTeamMembers
} from '../config/actions';
import teams from '../utilities/teams';
import user from '../utilities/user';
import safeURL from '../utilities/safeURL';
import notifications from '../utilities/notifications';

import SelectableBoxController from '../components/selectable-box/SelectableBoxController';
import Drawer from '../components/drawer/Drawer';
import TeamCreate from './team-create/TeamCreate';
import TeamDetails from './team-details/TeamDetails';
import TeamEditController from './team-edit/TeamEditController';
import TeamDeleteController from './team-delete/TeamDeleteController';
import Modal from '../components/Modal';

const propTypes = {
    dispatch: PropTypes.func.isRequired,
    allTeams: PropTypes.arrayOf(PropTypes.object).isRequired,
    activeTeam: PropTypes.object,
    rootPath: PropTypes.string.isRequired,
    routes: PropTypes.arrayOf(PropTypes.object).isRequired,
    userIsAdmin: PropTypes.bool.isRequired,
    params: PropTypes.object.isRequired
}

export class TeamsController extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isUpdatingAllTeams: false,
            isUpdatingTeamMembers: false,
            drawerIsAnimatable: false,
            clearActiveTeam: false,
            isEditingTeam: false,
            isDeletingTeam: false
        };

        this.handleTeamClick = this.handleTeamClick.bind(this);
        this.handleMembersEditClick = this.handleMembersEditClick.bind(this);
        this.handleDrawerTransitionEnd = this.handleDrawerTransitionEnd.bind(this);
        this.handleDrawerCancelClick = this.handleDrawerCancelClick.bind(this);
        this.handleTeamDeleteClick = this.handleTeamDeleteClick.bind(this);
        this.handleTeamDeleteSuccess = this.handleTeamDeleteSuccess.bind(this);
        this.handleTeamCreateSuccess = this.handleTeamCreateSuccess.bind(this);
    }

    componentWillMount() {
        this.fetchTeams();
    }

    componentWillReceiveProps(nextProps) {
        // Update with new active team
        const activeTeam = nextProps.allTeams.find(team => {
            return team.path === nextProps.params.team;
        });
        if (activeTeam && (nextProps.activeTeam.id !== activeTeam.id)) {
            this.fetchMembers(activeTeam.name);

            if (!this.props.params.team) {
                this.setState({drawerIsAnimatable: true});   
            }

            this.props.dispatch(updateActiveTeam(activeTeam));
            return;
        }
        
        // No active team in parameter anymore
        if (!nextProps.params.team && nextProps.activeTeam && nextProps.activeTeam.id) {
            this.setState({
                drawerIsAnimatable: true,
                clearActiveTeam: true
            });
        }

        // Open edit team modal
        if (nextProps.routes[nextProps.routes.length-1].path === "edit") {
            this.setState({isEditingTeam: true});
        }

        // Open delete team modal
        if (nextProps.routes[nextProps.routes.length-1].path === "delete") {
            this.setState({isDeletingTeam: true});
        }
    }

    shouldComponentUpdate(nextProps, nextState) {
        // Allow render because drawer animtation flag has changed
        if (this.state.drawerIsAnimatable !== nextState.drawerIsAnimatable) {
            return true;
        }

        // Allow render because the team editing modal is being displayed
        if (nextProps.routes[nextProps.routes.length-1].path === "edit" && nextState.isEditingTeam) {
            return true;
        }
        
        // Allow render because the team editing modal is being displayed
        if (nextProps.routes[nextProps.routes.length-1].path === "delete" && nextState.isDeletingTeam) {
            return true;
        }

        // Don't update component if all teams haven't been fetched yet
        if (!nextProps.allTeams) {
            return false;
        }

        // Component is still fetching teams and we don't have teams in Redux yet - don't render any changes
        if (nextState.isUpdatingAllTeams && (nextProps.allTeams.length === 0)) {
            return false;
        }

        return true;
    }

    componentWillUnmount() {
        this.props.dispatch(emptyActiveTeam());
    }

    handleDrawerTransitionEnd() {
        this.setState({drawerIsAnimatable: false});
        
        if (this.state.clearActiveTeam) {
            this.setState({clearActiveTeam: false});
            this.props.dispatch(emptyActiveTeam());
        }
    }

    handleMembersEditClick() {
        this.props.dispatch(push(`${this.props.rootPath}/teams/${this.props.activeTeam.path}/edit`));
    }

    handleDrawerCancelClick() {
        this.props.dispatch(push(`${this.props.rootPath}/teams`));
    }

    handleTeamDeleteClick() {
        this.props.dispatch(push(`${this.props.rootPath}/teams/${this.props.activeTeam.path}/delete`));
    }

    handleTeamDeleteSuccess() {
        const notification = {
            type: 'positive',
            message: `Team '${this.props.activeTeam.name}' successfully deleted`,
            isDismissable: true,
            autoDismiss: 15000
        }
        notifications.add(notification);
        this.fetchTeams();
    }

    handleTeamCreateSuccess() {
        this.fetchTeams();
    }

    fetchTeams() {
        this.setState({isUpdatingAllTeams: true});
        teams.getAll().then(allTeams => {
            // Add any props (such as 'path') to response from API
            const allTeamsWithProps = allTeams.map(team => {
                const path = safeURL(team.name + "_" + team.id);
                return Object.assign({}, team, {
                    path: path
                });
            });
            
            // Update all teams
            const teamParameter = this.props.params.team;
            this.props.dispatch(updateAllTeams(allTeamsWithProps));
            this.setState({isUpdatingAllTeams: false});

            // Update active team
            if (teamParameter) {
                const activeTeam = allTeamsWithProps.find(team => {
                    return team.path === teamParameter;
                });
                // Only update Redux if new active team is different from the current one
                if (activeTeam && activeTeam !== this.props.activeTeam ) {
                    this.props.dispatch(updateActiveTeam(activeTeam));
                    return;
                }
                // Give error because the team in the URL can't be found in the data
                if (!activeTeam) {
                    const notification = {
                        message: `Team '${teamParameter}' is not recognised so you've been redirected to the teams screen`,
                        type: "neutral",
                        autoDismiss: 15000,
                        isDismissable: true
                    }
                    notifications.add(notification);
                    this.props.dispatch(push(`${this.props.rootPath}/teams`));
                }
            }
        }).catch(error => {
            switch(error.status) {
                // TODO any 401s on http request could be handled globally
                case(401): {
                    const notification = {
                        type: "neutral",
                        message: "Your session has expired so you've been redirected to the login screen",
                        isDismissable: true,
                        autoDismiss: 20000
                    }
                    user.logOut();
                    this.props.dispatch(push(`${this.props.rootPath}/login?redirect=${this.props.rootPath}/teams`));
                    notifications.add(notification);
                    break;
                }
                case("RESPONSE_ERR"): {
                    const notification = {
                        type: "warning",
                        message: "An error's occurred whilst trying to get teams. You may only be able to see previously loaded information but won't be able to edit any team members",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
                case("UNEXPECTED_ERR"): {
                    const notification = {
                        type: "warning",
                        message: "An unexpected error's occurred whilst trying to get teams. You may only be able to see previously loaded information but won't be able to edit any team members",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
                case("FETCH_ERR"): {
                    const notification = {
                        type: "warning",
                        message: "There's been a network error whilst trying to get teams. You may only be able to see previously loaded information and not be able to edit any team members",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
            }

            console.error("Error fetching all teams:\n", error);
        });
    }

    fetchMembers(teamName) {
        this.setState({isUpdatingTeamMembers: true});
        teams.get(teamName).then(team => {
            // A new team is now active, don't do anything with the fetched data
            if (teamName !== this.props.activeTeam.name) {
                return;
            }
            
            this.props.dispatch(updateActiveTeamMembers(team.members));
            this.setState({isUpdatingTeamMembers: false});
        }).catch(error => {
            switch(error.status) {
                // TODO any 401s on http request could be handled globally
                case(401): {
                    const notification = {
                        type: "neutral",
                        message: "Your session has expired so you've been redirected to the login screen",
                        isDismissable: true,
                        autoDismiss: 20000
                    }
                    user.logOut();
                    this.props.dispatch(push(`${this.props.rootPath}/login?redirect=${this.props.rootPath}/teams/${this.props.activeTeam.path}`));
                    notifications.add(notification);
                    break;
                }
                case("RESPONSE_ERR"): {
                    const notification = {
                        type: "warning",
                        message: `An error's occurred whilst trying to get the members for the team '${teamName}'`,
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
                case("UNEXPECTED_ERR"): {
                    const notification = {
                        type: "warning",
                        message: `An unexpected error's occurred whilst trying to get the members for the team '${teamName}'`,
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
                case("FETCH_ERR"): {
                    const notification = {
                        type: "warning",
                        message: `There's been a network error whilst trying to get the members for the team '${teamName}'`,
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
            }
            console.error(`Error fetching team '${teamName}':\n`, error);
        });
    }

    handleTeamClick(clickedTeam) {
        // Make no change if clicked team is already the selected team
        if (clickedTeam.isSelected) {
            return;
        }
        const path = safeURL(clickedTeam.name + "_" + clickedTeam.id);
        this.props.dispatch(push(`${this.props.rootPath}/teams/${path}`));
    }

    renderDrawer() {
        return (
            <Drawer
                isVisible={this.props.activeTeam && this.props.activeTeam.id && !this.state.clearActiveTeam ? true : false} 
                isAnimatable={this.state.drawerIsAnimatable} 
                handleTransitionEnd={this.handleDrawerTransitionEnd}
            >
                {
                    this.props.activeTeam && this.props.activeTeam.id ?
                        <TeamDetails 
                            {...this.props.activeTeam} 
                            userIsAdmin={this.props.userIsAdmin} 
                            onCancel={this.handleDrawerCancelClick}
                            onDelete={this.handleTeamDeleteClick}
                            onEditMembers={this.handleMembersEditClick}
                            isShowingLoader={this.state.isUpdatingTeamMembers}
                            isReadOnly={this.state.isUpdatingAllTeams}
                        />
                        :
                        ""
                }
            </Drawer>
        )
    }

    render() {
        return (
            <div>
                <div className="grid grid--justify-space-around">
                    <div className="grid__col-4">
                        <h1>Select a team</h1>
                        <SelectableBoxController 
                            items={this.props.allTeams}
                            activeItem={this.props.activeTeam}
                            isUpdating={this.state.isUpdatingAllTeams} 
                            heading="Name"
                            handleItemClick={this.handleTeamClick}
                        />
                    </div>
                    <div className="grid__col-4">
                        <h1>Create a team</h1>
                        <TeamCreate onCreateSuccess={this.handleTeamCreateSuccess} />
                    </div>
                </div>
                {this.renderDrawer()}
                {
                    this.props.routes[this.props.routes.length-1].path === "edit" && this.props.activeTeam && this.props.activeTeam.id ?
                    <Modal sizeClass="grid__col-8">
                        <TeamEditController isUpdatingMembers={this.state.isUpdatingTeamMembers}/>
                    </Modal>
                    :
                    ""
                }
                {
                    this.props.routes[this.props.routes.length-1].path === "delete" && this.props.activeTeam && this.props.activeTeam.id ?
                    <Modal sizeClass="grid__col-3">
                        <TeamDeleteController name={this.props.activeTeam.name} onDeleteSuccess={this.handleTeamDeleteSuccess}/>
                    </Modal>
                    :
                    ""
                }
            </div>
        )
    }
}

TeamsController.propTypes = propTypes;

function mapStateToProps(state) {
    return {
        activeTeam: state.state.teams.active,
        allTeams: state.state.teams.all,
        userIsAdmin: state.state.user.isAdmin,
        rootPath: state.state.rootPath
    }
}

export default connect(mapStateToProps)(TeamsController);