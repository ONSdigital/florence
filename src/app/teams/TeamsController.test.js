import React from 'react';
import { TeamsController } from './TeamsController';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

const listOfTeams = [
    {
        name: "Team 1",
        id: 1,
        path: "team_1_1"
    },
    {
        name: "Team 2",
        id: 2,
        path: "team_2_2"
    },
    {
        name: "Team 2",
        id: 3,
        path: "team_3_3"
    }
]

jest.mock('../utilities/teams.js', () => (
    {
        getAll: jest.fn().mockImplementation(() => {
            return Promise.resolve([
                {
                    name: "Team 1",
                    id: 1,
                    path: "team_1_1"
                },
                {
                    name: "Team 2",
                    id: 2,
                    path: "team_2_2"
                },
                {
                    name: "Team 2",
                    id: 3,
                    path: "team_3_3"
                }
            ]);
        }
    )}
));

test('Loading state shown when fetching all teams', () => {
    const props = {
        dispatch: function() {},
        rootPath: '/florence',
        allTeams: listOfTeams,
        params: {},
        userIsAdmin: false,
        routes: [
            {
                path: "florence"
            },
            {
                path: "teams"
            }
        ]
    }
    const component = mount(
        <TeamsController {...props}/>
    )
    expect(component.find('.loader').length).toBe(1);
    component.setState({isUpdatingAllTeams: false});
    expect(component.find('.loader').length).toBe(0);
});

test('Renders updated list of teams', () => {
    const props = {
        dispatch: function() {},
        rootPath: '/florence',
        allTeams: [],
        params: {},
        userIsAdmin: false,
        routes: [
            {
                path: "florence"
            },
            {
                path: "teams"
            }
        ]
    }
    const component = mount(
        <TeamsController {...props}/>
    )
    expect(component.find('li').length).toBe(0);
    component.setProps({allTeams: listOfTeams});
    component.setState({isUpdatingAllTeams: false});
    expect(component.find('li').length).toBe(3);
});

test('Correctly renders when the active team is changed', () => {
    const props = {
        dispatch: function() {},
        rootPath: '/florence',
        allTeams: listOfTeams,
        params: {},
        activeTeam: {},
        userIsAdmin: false,
        routes: [
            {
                path: "florence"
            },
            {
                path: "teams"
            }
        ]
    }
    const component = mount(
        <TeamsController {...props}/>
    )
    component.setState({isUpdatingAllTeams: false});
    expect(component.find('.selected').length).toBe(0);
    component.setProps({
        activeTeam: {
            name: "Team 1",
            id: 1,
            members: ["tester 1", "tester 2"]
        }
    });
    expect(component.find('.selected').length).toBe(1);
    component.setProps({activeTeam: {}});
    expect(component.find('.selected').length).toBe(0);
});

test("Non-admin users can't view option to edit teams", () => {
    const props = {
        dispatch: function() {},
        rootPath: '/florence',
        allTeams: listOfTeams,
        params: {
            team: "team_1_1"
        },
        activeTeam: {
            name: "Team 1",
            id: 1,
            members: ["tester 1", "tester 2"]
        },
        userIsAdmin: false,
        routes: [
            {
                path: "florence"
            },
            {
                path: "teams"
            }
        ]
    }
    const component = renderer.create(
        <TeamsController {...props} />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test("Admin users have option to edit teams", () => {
    const props = {
        dispatch: function() {},
        rootPath: '/florence',
        allTeams: listOfTeams,
        params: {
            team: "team_1_1"
        },
        activeTeam: {
            name: "Team 1",
            id: 1,
            members: ["tester 1", "tester 2"]
        },
        userIsAdmin: true,
        routes: [
            {
                path: "florence"
            },
            {
                path: "teams"
            }
        ]
    }
    const component = renderer.create(
        <TeamsController {...props} />
    );
    expect(component.toJSON()).toMatchSnapshot();
});