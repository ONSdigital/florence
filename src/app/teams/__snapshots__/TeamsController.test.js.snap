// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin users have option to edit teams 1`] = `
<div>
  <div
    className="grid grid--justify-space-around"
  >
    <div
      className="grid__col-4"
    >
      <h1>
        Select a team
      </h1>
      <div
        className="selectable-box"
      >
        <h2
          className="selectable-box__heading"
        >
          Name
          <span
            className="selectable-box__status loader"
          />
        </h2>
        <ul
          className="selectable-box__list"
        >
          <li
            className="selectable-box__item selected"
            id={1}
            onClick={[Function]}
          >
            Team 1
          </li>
          <li
            className="selectable-box__item "
            id={2}
            onClick={[Function]}
          >
            Team 2
          </li>
          <li
            className="selectable-box__item "
            id={3}
            onClick={[Function]}
          >
            Team 2
          </li>
        </ul>
      </div>
    </div>
    <div
      className="grid__col-4"
    >
      <h1>
        Create a team
      </h1>
      <form
        className="form "
        onSubmit={[Function]}
      >
        <div
          className="form__input"
        >
          <input
            className="input input__text"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
          
        </div>
        <button
          className="btn btn--positive"
        >
          Create
        </button>
      </form>
    </div>
  </div>
  <div
    className="drawer visible"
    onTransitionEnd={[Function]}
  >
    <div
      className="drawer__container"
    >
      <h2
        className="drawer__heading"
      >
        Team 1
      </h2>
      <div
        className="drawer__banner drawer__banner--dark"
      >
        <h3>
          Read only
        </h3>
        <p>
          The list of teams if still being fetched so you can't edit or delete a team
        </p>
      </div>
      <div
        className="drawer__banner"
      >
        <button
          className="btn btn--primary"
          disabled={true}
          onClick={[Function]}
        >
          Add / remove members
        </button>
      </div>
      <div
        className="drawer__body"
      >
        <ul
          className="list list--neutral"
        >
          <li
            className="list__item"
          >
            tester 1
          </li>
          <li
            className="list__item"
          >
            tester 2
          </li>
        </ul>
      </div>
      <div
        className="drawer__footer"
      >
        <button
          className="btn btn--warning btn--margin-right"
          disabled={true}
          onClick={[Function]}
        >
          Delete team
        </button>
        <button
          className="btn"
          onClick={[Function]}
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
  
  
</div>
`;

exports[`Non-admin users can't view option to edit teams 1`] = `
<div>
  <div
    className="grid grid--justify-space-around"
  >
    <div
      className="grid__col-4"
    >
      <h1>
        Select a team
      </h1>
      <div
        className="selectable-box"
      >
        <h2
          className="selectable-box__heading"
        >
          Name
          <span
            className="selectable-box__status loader"
          />
        </h2>
        <ul
          className="selectable-box__list"
        >
          <li
            className="selectable-box__item selected"
            id={1}
            onClick={[Function]}
          >
            Team 1
          </li>
          <li
            className="selectable-box__item "
            id={2}
            onClick={[Function]}
          >
            Team 2
          </li>
          <li
            className="selectable-box__item "
            id={3}
            onClick={[Function]}
          >
            Team 2
          </li>
        </ul>
      </div>
    </div>
    <div
      className="grid__col-4"
    >
      <h1>
        Create a team
      </h1>
      <form
        className="form "
        onSubmit={[Function]}
      >
        <div
          className="form__input"
        >
          <input
            className="input input__text"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
          
        </div>
        <button
          className="btn btn--positive"
        >
          Create
        </button>
      </form>
    </div>
  </div>
  <div
    className="drawer visible"
    onTransitionEnd={[Function]}
  >
    <div
      className="drawer__container"
    >
      <h2
        className="drawer__heading"
      >
        Team 1
      </h2>
      <div
        className="drawer__banner drawer__banner--dark"
      >
        <h3>
          Read only
        </h3>
        <p>
          The list of teams if still being fetched so you can't edit or delete a team
        </p>
      </div>
      <div
        className="drawer__body"
      >
        <ul
          className="list list--neutral"
        >
          <li
            className="list__item"
          >
            tester 1
          </li>
          <li
            className="list__item"
          >
            tester 2
          </li>
        </ul>
      </div>
      <div
        className="drawer__footer"
      >
        <button
          className="btn btn--warning btn--margin-right"
          disabled={true}
          onClick={[Function]}
        >
          Delete team
        </button>
        <button
          className="btn"
          onClick={[Function]}
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
  
  
</div>
`;
